File: backend/logger.py
"""
Sistema de logging centralizado para el backend de SARA.
Captura toda la salida de consola con fecha y hora.
"""

import sys
import os
from datetime import datetime
from pathlib import Path


class Logger:
    """Logger personalizado que captura stdout y stderr con timestamps."""
    
    def __init__(self, log_dir="logs", log_file_prefix="sara_backend"):
        self.log_dir = Path(log_dir)
        self.log_dir.mkdir(exist_ok=True)
        
        # Crear archivo de log con fecha
        timestamp = datetime.now().strftime("%Y%m%d")
        self.log_file = self.log_dir / f"{log_file_prefix}_{timestamp}.log"
        
        # Guardar los streams originales
        self.original_stdout = sys.stdout
        self.original_stderr = sys.stderr
        
        # Inicializar el logger
        self._setup_logging()
    
    def _get_timestamp(self):
        """Genera timestamp con fecha y hora."""
        return datetime.now().strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]
    
    def _write_log(self, message, stream_type="INFO"):
        """Escribe un mensaje al archivo de log con timestamp."""
        timestamp = self._get_timestamp()
        log_line = f"[{timestamp}] [{stream_type}] {message}\n"
        
        # Escribir al archivo
        try:
            with open(self.log_file, 'a', encoding='utf-8') as f:
                f.write(log_line)
        except Exception as e:
            # Si falla escribir al archivo, al menos imprimir al stream original
            self.original_stdout.write(f"Error writing to log: {e}\n")
    
    def _create_stream_wrapper(self, original_stream, stream_type):
        """Crea un wrapper que intercepta la salida del stream."""
        class StreamWrapper:
            def __init__(self, logger_instance, orig_stream, type_name):
                self.logger = logger_instance
                self.orig_stream = orig_stream
                self.type_name = type_name
                self.buffer = ""
            
            def write(self, text):
                # Escribir al stream original
                self.orig_stream.write(text)
                
                # Acumular en buffer para líneas completas
                self.buffer += text
                
                # Procesar líneas completas
                while '\n' in self.buffer:
                    line, self.buffer = self.buffer.split('\n', 1)
                    if line:  # Solo loggear si hay contenido
                        self.logger._write_log(line, self.type_name)
                
                # Forzar flush del stream original
                self.orig_stream.flush()
            
            def flush(self):
                # Procesar cualquier contenido restante en el buffer
                if self.buffer:
                    self.logger._write_log(self.buffer, self.type_name)
                    self.buffer = ""
                self.orig_stream.flush()
        
        return StreamWrapper(self, original_stream, stream_type)
    
    def _setup_logging(self):
        """Configura el sistema de logging interceptando stdout y stderr."""
        # Crear wrappers para stdout y stderr
        sys.stdout = self._create_stream_wrapper(self.original_stdout, "STDOUT")
        sys.stderr = self._create_stream_wrapper(self.original_stderr, "STDERR")
        
        # Log inicial
        self._write_log("=" * 80, "INFO")
        self._write_log("Logger inicializado. Capturando stdout y stderr.", "INFO")
        self._write_log(f"Archivo de log: {self.log_file}", "INFO")
        self._write_log("=" * 80, "INFO")
    
    def info(self, message):
        """Log un mensaje de info explícito."""
        self._write_log(message, "INFO")
    
    def error(self, message):
        """Log un mensaje de error explícito."""
        self._write_log(message, "ERROR")
    
    def warning(self, message):
        """Log un mensaje de advertencia explícito."""
        self._write_log(message, "WARNING")
    
    def debug(self, message):
        """Log un mensaje de debug explícito."""
        self._write_log(message, "DEBUG")
    
    def restore_original_streams(self):
        """Restaura los streams originales de stdout y stderr."""
        sys.stdout = self.original_stdout
        sys.stderr = self.original_stderr
        self._write_log("Streams originales restaurados.", "INFO")


# Instancia global del logger
_logger_instance = None


def init_logger(log_dir="logs", log_file_prefix="sara_backend"):
    """Inicializa el logger global."""
    global _logger_instance
    if _logger_instance is None:
        _logger_instance = Logger(log_dir, log_file_prefix)
    return _logger_instance


def get_logger():
    """Obtiene la instancia del logger global."""
    return _logger_instance


def log(message, level="INFO"):
    """Función conveniente para loggear mensajes."""
    if _logger_instance:
        _logger_instance._write_log(message, level)

File: electron/logger.js
/**
 * Sistema de logging para el proceso principal de Electron.
 * Captura toda la salida de consola con fecha y hora.
 */

const fs = require('fs');
const path = require('path');

class ElectronLogger {
    constructor(logDir = 'logs', logFilePrefix = 'ada_electron') {
        this.logDir = path.join(__dirname, '..', logDir);
        this.logFilePrefix = logFilePrefix;
        
        // Crear directorio de logs si no existe
        if (!fs.existsSync(this.logDir)) {
            fs.mkdirSync(this.logDir, { recursive: true });
        }
        
        // Generar nombre de archivo con fecha
        const date = new Date();
        const dateStr = date.toISOString().split('T')[0].replace(/-/g, '');
        this.logFile = path.join(this.logDir, `${logFilePrefix}_${dateStr}.log`);
        
        // Guardar las funciones originales de console
        this.originalConsole = {
            log: console.log,
            error: console.error,
            warn: console.warn,
            info: console.info,
            debug: console.debug
        };
        
        // Inicializar el logger
        this._setupLogging();
    }
    
    _getTimestamp() {
        const now = new Date();
        const year = now.getFullYear();
        const month = String(now.getMonth() + 1).padStart(2, '0');
        const day = String(now.getDate()).padStart(2, '0');
        const hours = String(now.getHours()).padStart(2, '0');
        const minutes = String(now.getMinutes()).padStart(2, '0');
        const seconds = String(now.getSeconds()).padStart(2, '0');
        const ms = String(now.getMilliseconds()).padStart(3, '0');
        return `${year}-${month}-${day} ${hours}:${minutes}:${seconds}.${ms}`;
    }
    
    _writeLog(message, level = 'INFO') {
        const timestamp = this._getTimestamp();
        const logLine = `[${timestamp}] [${level}] ${message}\n`;
        
        try {
            fs.appendFileSync(this.logFile, logLine, 'utf-8');
        } catch (error) {
            // Si falla escribir al archivo, al menos imprimir a la consola original
            this.originalConsole.error(`Error writing to log: ${error.message}`);
        }
    }
    
    _setupLogging() {
        // Log inicial
        this._writeLog('='.repeat(80), 'INFO');
        this._writeLog('Logger inicializado. Capturando console.log, console.error, etc.', 'INFO');
        this._writeLog(`Archivo de log: ${this.logFile}`, 'INFO');
        this._writeLog('='.repeat(80), 'INFO');
        
        // Sobrescribir console.log
        console.log = (...args) => {
            const message = args.map(arg => 
                typeof arg === 'object' ? JSON.stringify(arg, null, 2) : String(arg)
            ).join(' ');
            this.originalConsole.log(...args);
            this._writeLog(message, 'STDOUT');
        };
        
        // Sobrescribir console.error
        console.error = (...args) => {
            const message = args.map(arg => 
                typeof arg === 'object' ? JSON.stringify(arg, null, 2) : String(arg)
            ).join(' ');
            this.originalConsole.error(...args);
            this._writeLog(message, 'STDERR');
        };
        
        // Sobrescribir console.warn
        console.warn = (...args) => {
            const message = args.map(arg => 
                typeof arg === 'object' ? JSON.stringify(arg, null, 2) : String(arg)
            ).join(' ');
            this.originalConsole.warn(...args);
            this._writeLog(message, 'WARNING');
        };
        
        // Sobrescribir console.info
        console.info = (...args) => {
            const message = args.map(arg => 
                typeof arg === 'object' ? JSON.stringify(arg, null, 2) : String(arg)
            ).join(' ');
            this.originalConsole.info(...args);
            this._writeLog(message, 'INFO');
        };
        
        // Sobrescribir console.debug
        console.debug = (...args) => {
            const message = args.map(arg => 
                typeof arg === 'object' ? JSON.stringify(arg, null, 2) : String(arg)
            ).join(' ');
            this.originalConsole.debug(...args);
            this._writeLog(message, 'DEBUG');
        };
    }
    
    info(message) {
        this._writeLog(message, 'INFO');
    }
    
    error(message) {
        this._writeLog(message, 'ERROR');
    }
    
    warning(message) {
        this._writeLog(message, 'WARNING');
    }
    
    debug(message) {
        this._writeLog(message, 'DEBUG');
    }
    
    restoreOriginalConsole() {
        console.log = this.originalConsole.log;
        console.error = this.originalConsole.error;
        console.warn = this.originalConsole.warn;
        console.info = this.originalConsole.info;
        console.debug = this.originalConsole.debug;
        this._writeLog('Consola original restaurada.', 'INFO');
    }
}

// Instancia global del logger
let _loggerInstance = null;

function initLogger(logDir = 'logs', logFilePrefix = 'ada_electron') {
    if (_loggerInstance === null) {
        _loggerInstance = new ElectronLogger(logDir, logFilePrefix);
    }
    return _loggerInstance;
}

function getLogger() {
    return _loggerInstance;
}

module.exports = {
    ElectronLogger,
    initLogger,
    getLogger
};
